src/vec.rs

```diff
@@ -0,0 +1,364 @@
+        v
+    }
+
+    pub fn load(memory: M) -> Self {
+        Self::load_with_sizes(memory, T::MAX_SIZE as u64)
+    }
+
+    pub fn load_with_sizes(memory: M, max_value_size: u64) -> Self {
+        // Read the header from memory.
+        let header: StableVecHeader = read_struct(Address::from(0), &memory);
+        assert_eq!(&header.magic, MAGIC, "Bad magic.");
+        assert_eq!(header.version, LAYOUT_VERSION, "Unsupported version.");
+        assert_eq!(header.fixed_value_size, T::FIXED_SIZE);
+        let expected_value_size = header.max_value_size;
+        assert!(
+            max_value_size <= expected_value_size,
+            "max_value_size must be <= {}",
+            expected_value_size
+        );
+
+        Self {
+            memory,
+            max_value_size: header.max_value_size,
+            base: Address::from(0) + StableVecHeader::size(),
+            len: header.len,
+            _phantom: PhantomData,
+        }
+    }
+
+    fn save(&self) {
+        let header = StableVecHeader {
+            magic: *MAGIC,
+            version: LAYOUT_VERSION,
+            max_value_size: self.max_value_size,
+            fixed_value_size: T::FIXED_SIZE,
+            len: self.len,
+            _buffer: [0; 24],
+        };
+
+        write_struct(&header, Address::from(0), &self.memory);
+    }
+
+    pub fn clear(&mut self) {
+        self.len = 0;
+        self.save();
+    }
+
+    #[inline]
+    fn index_to_entry_addr(&self, index: usize) -> Address {
+        self.base + Bytes::from((Self::ENTRY_SIZE * index) as u64)
+    }
+
+    #[inline]
+    fn read_value(&self, index: usize) -> T {
+        let mut p = self.index_to_entry_addr(index);
+        // ret = read_struct(ptr, &self.memory);
+        let mut buf = if T::FIXED_SIZE {
+            vec![0; T::MAX_SIZE as usize]
+        } else {
+            let size = read_u32(&self.memory, p) as usize;
+            p += Bytes::from(4u64);
+            vec![0; size]
+        };
+        self.memory.read(p.get(), &mut buf);
+        T::from_bytes(buf)
+    }
+
+    #[inline]
+    fn write_value(&mut self, index: usize, value: &T) {
+        let value = value.to_bytes();
+        let mut p = self.index_to_entry_addr(index);
+        if !T::FIXED_SIZE {
+            write_u32(&self.memory, p, value.len() as u32);
+            p += Bytes::from(4u64);
+        }
+        self.memory.write(p.get(), &value);
+    }
+
+    pub fn insert(&mut self, index: usize, element: &T) {
+        let len = self.len();
+        assert!(index <= len);
+
+        // space for the new element
+        if len == self.capacity() {
+            self.reserve(1);
+        }
+
+        // Move all entries starting at index one to the right.
+        let src = self.index_to_entry_addr(index);
+        let dst = self.index_to_entry_addr(index + 1);
+        copy(
+            &self.memory,
+            src,
+            dst,
+            (len - index) as u64,
+            Self::ENTRY_SIZE,
+        );
```

Comment:
The amount of memory that needs to be shifted here is unbounded. If I have a vector containing millions of entries, this call may need to move gigabytes of state. We have a limit on how much computation we can do per message (measured in a unit called "instructions"), and this kind of shifting with a large enough state would cause whatever is currently running to panic.

Reply:
see https://github.com/dfinity/stable-structures/pull/18/files/970473e8601434c19f146d7cbf341a52750eeee1#r1033744554